### PR TITLE
RavenDB-22076 Implementing HNSW

### PR DESCRIPTION
**Pull request by ayende**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22076 

Implementing HNSW approximate nearest neighbor

### CODE REVIEW

File: src/Voron/Data/Graphs/Hnsw.SimilarityMethods.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public enum SimilarityMethod : byte
+    {
+        CosineSimilaritySingles = 0,
+        CosineSimilarityI8 = 1,
+        HammingDistance = 2,
+    }
+    
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static float CosineSimilaritySingles(Span<byte> a, Span<byte> b) => TensorPrimitives.CosineSimilarity(MemoryMarshal.Cast<byte, float>(a), MemoryMarshal.Cast<byte, float>(b));
```
Comment (ayende) [inline src/Voron/Data/Graphs/Hnsw.SimilarityMethods.cs]: Shouldn't this be?

```suggestion
    private static float CosineSimilaritySingles(Span<byte> a, Span<byte> b) => 1- TensorPrimitives.CosineSimilarity(MemoryMarshal.Cast<byte, float>(a), MemoryMarshal.Cast<byte, float>(b));
```

File: src/Voron/Data/Graphs/Hnsw.cs [hunk 2]
```diff
@@ -479,10 +470,7 @@ public float Distance(Span<byte> vector, int fromIdx, int toIdx)
 
             ref var to = ref GetNodeByIndex(toIdx);
             Span<byte> v2 = to.GetVector(in this);
-            var distance = 1 - TensorPrimitives.CosineSimilarity(
-                MemoryMarshal.Cast<byte, float>(vector),
-                MemoryMarshal.Cast<byte, float>(v2)
-            );
+            var distance = 1 - SimilarityCalc(vector, v2);
```
Comment (ayende) [inline src/Voron/Data/Graphs/Hnsw.cs]: You have `1 - ` in the `HammingDistance`, however.
Better to be consistence and have every `1 - ` etc in the actual functions, not here

Comment (maciejaszyk) [reply src/Voron/Data/Graphs/Hnsw.cs]: Done

File: src/Voron/Data/Graphs/Hnsw.SimilarityMethods.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Voron/Data/Graphs/Hnsw.SimilarityMethods.cs]: Done

**Comment by maciejaszyk** [comment]
This is part of https://github.com/ravendb/ravendb/pull/19431

**Pull request closed without merging**